zapcore/sampler.go

```diff
@@ -126,9 +183,49 @@ func (s *sampler) Check(ent Entry, ce *CheckedEntry) *CheckedEntry {
 	}
 
 	counter := s.counts.get(ent.Level, ent.Message)
-	n := counter.IncCheckReset(ent.Time, s.tick)
+	n, prevN := counter.IncCheckReset(ent.Time, s.tick)
 	if n > s.first && (n-s.first)%s.thereafter != 0 {
 		return ce
 	}
+	if err := s.Reporter.Report(ent, s.getSampleCount(prevN)); err != nil {
```

Comment:
not sure if we can make a change here, but it's a little unexpected that `Check` does a write here.

should we instead add hooks into `CheckedEntry` that trigger a function to run on `Write`?

Reply:
Hmm, but wouldn't that only write during the next unsampled log entry, which could be much later as you point out? Maybe that's okay? Again it all comes down to: what actually is the goal here.

At any rate, `Check` doing a write is mildly less surprising than `With` doing a write as was the initial version.

Reply:
If we wanted to be aware that sampling was happening (e.g., log as soon as we know we'll sample), then I think we could make it work by either:

 * Having the log that gets to the sampling threshold also log "all future entries will be sampled". The con is that we'll log this even if there were no other sampled logs. It could get noisy for low sampling rates (like 1 a second) even if the event only happens once.
 * Having the first log that would be sampled returning a checked entry, but the write doesn't write the original log, but a "dropping log due to sample". This might be surprising since the entry that is written isn't the entry that was checked.

There could also be a slightly different approach, where the original log entry has an added field like "sampleLimitHit" or something, which makes the first option less intrusive.